pkg/controller/plan/migrator/base/migrator.go

```diff
@@ -0,0 +1,318 @@
+			default:
+				err = liberr.New(fmt.Sprintf("Unknown step '%s'. Not implemented.", step.Name))
+				return
+			}
+			pipeline = append(
+				pipeline,
+				&plan.Step{
+					Task: plan.Task{
+						Name:        taskName,
+						Description: taskDescription,
+						Progress: libitr.Progress{
+							Total: total,
+						},
+						Annotations: map[string]string{
+							"unit": "MB",
+						},
+						Phase: Pending,
+					},
+					Tasks: tasks,
+				})
+		case Finalize:
+			tasks, pErr := r.builder.Tasks(vm.Ref)
+			if pErr != nil {
+				err = liberr.Wrap(pErr)
+				return
+			}
+			total := int64(0)
+			for _, task := range tasks {
+				total += task.Progress.Total
+			}
+			pipeline = append(
+				pipeline,
+				&plan.Step{
+					Task: plan.Task{
+						Name:        Cutover,
+						Description: "Finalize disk transfer.",
+						Progress: libitr.Progress{
+							Total: total,
+						},
+						Annotations: map[string]string{
+							"unit": "MB",
+						},
+					},
+					Tasks: tasks,
+				})
+		case ConvertGuest:
+			pipeline = append(
+				pipeline,
+				&plan.Step{
+					Task: plan.Task{
+						Name:        ImageConversion,
+						Description: "Convert image to kubevirt.",
+						Progress:    libitr.Progress{Total: 1},
+						Phase:       Pending,
+					},
+				})
+		case PostHook:
+			pipeline = append(
+				pipeline,
+				&plan.Step{
+					Task: plan.Task{
+						Name:        PostHook,
+						Description: "Run post-migration hook.",
+						Progress:    libitr.Progress{Total: 1},
+						Phase:       Pending,
+					},
+				})
+		case CreateVM:
+			pipeline = append(
+				pipeline,
+				&plan.Step{
+					Task: plan.Task{
+						Name:        VMCreation,
+						Description: "Create VM.",
+						Phase:       Pending,
+						Progress:    libitr.Progress{Total: 1},
+					},
+				})
+		}
+		next, done, _ := itinerary.Next(step.Name)
+		if !done {
+			step = next
+		} else {
+			break
+		}
+	}
+
+	if len(pipeline) == 0 {
+		err = liberr.New("Empty pipeline.", "vm", vm.String())
+		return
+	}
+
+	r.Log.V(2).Info(
+		"Pipeline built.",
+		"vm",
+		vm.String())
+	return
+}
+
+func (r *BaseMigrator) Itinerary(predicate libitr.Predicate) (itinerary libitr.Itinerary) {
+	if r.Context.Plan.Spec.Warm {
+		itinerary = WarmItinerary
+	} else {
+		itinerary = ColdItinerary
+	}
+	itinerary.Predicate = predicate
+	return
+}
+
+func (r *BaseMigrator) Next(status *plan.VMStatus) (next string) {
+	itinerary := r.Itinerary(&BasePredicate{vm: &status.VM, context: r.Context})
+	step, done, err := itinerary.Next(status.Phase)
+	if done || err != nil {
+		next = Completed
+		if err != nil {
+			log.Error(err, "Next phase failed.")
+		}
+	} else {
+		next = step.Name
+	}
+	r.Log.Info("Itinerary transition", "current phase", status.Phase, "next phase", next)
+	return
+}
+
+func (r *BaseMigrator) ExecutePhase(vm *plan.VMStatus) (ok bool, err error) {
+	// return ok = false to delegate to default itinerary implementation
+	// in plan/migration.go
+	return
```

Comment:
Is this for future usage?

Reply:
Yes. This is an extension point for a Migrator to provide its own implementation of a migration phase. Eventually the shared implementation in the migration runner in controller/plan/migration.go:execute() should go away and phase execution should be done by the provider migrators. In order to avoid requiring the entire system to be refactored at once, the BaseMigrator in this module just hands control back to the shared implementation.

(This way refactoring can be done in stages, first with the shared implementation being moved into this BaseMigrator, and then eventually broken down into vsphere, ovirt, ova, etc packages.)